File: solutions/java/code-carl/ch09_dp/T12_LC0474_FindMaxForm.java
/**
 * Practice template for:
 * - LeetCode 474. Ones and Zeroes
 *   https://leetcode.cn/problems/ones-and-zeroes/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/22 16:52
 * @comment
 */
public class T12_LC0474_FindMaxForm {
    public int findMaxForm(String[] strs, int m, int n) {
        int[][] dp = new int[m + 1][n + 1];
        for (String str : strs) {
            int count0 = 0, count1 = 0;
            for (int k = 0; k < str.length(); k++) {
                if (str.charAt(k) == '0') {
                    count0++;
                } else if (str.charAt(k) == '1') {
                    count1++;
                }
            }
            for (int i = m; i >= count0; i--) {
                for (int j = n; j >= count1; j--) {
                    dp[i][j] = Math.max(dp[i][j], dp[i - count0][j - count1] + 1);
                }
            }
        }
        return dp[m][n];
    }
}

File: solutions/python/code-carl/ch09_dp/t12_lc0474_find_max_form.py
"""LeetCode 474. Ones and Zeroes
https://leetcode.cn/problems/ones-and-zeroes/
"""
from collections import Counter
from typing import List


class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        dp: List[List[int]] = [[0] * (n + 1) for _ in range(m + 1)]
        for s in strs:
            cnt = Counter(s)
            for i in range(m, cnt['0'] - 1, -1):
                for j in range(n, cnt['1'] - 1, -1):
                    dp[i][j] = max(dp[i][j], dp[i - cnt['0']][j - cnt['1']] + 1)

        return dp[-1][-1]


if __name__ == "__main__":
    solution = Solution()
    assert solution.findMaxForm(['10', '0001', '111001', '1', '0'], 5, 3) == 4
    assert solution.findMaxForm(['10', '0', '1'], 1, 1) == 2
    print("All tests passed.")
